pkg/kv/kvtest/store.go

```diff
@@ -0,0 +1,269 @@
+	defer store.Close()
+
+	testKey := []byte("key")
+	testValue1 := []byte("value1")
+	testValue2 := []byte("value2")
+
+	// set test key with value1
+	err := store.Set(ctx, testKey, testValue1)
+	if err != nil {
+		t.Fatalf("failed to set key '%s', to value '%s': %s", testKey, testValue1, err)
+	}
+
+	// get test key with value1
+	val, err := store.Get(ctx, testKey)
+	switch {
+	case err != nil:
+		t.Fatalf("failed to get key '%s': %s", testKey, err)
+	case val == nil:
+		t.Fatalf("got value with nil")
+	case !bytes.Equal(testValue1, val):
+		t.Fatalf("key='%s' value='%s' doesn't match, expected='%s'", testKey, val, testValue1)
+	}
+
+	// override key with value2
+	err = store.Set(ctx, testKey, testValue2)
+	if err != nil {
+		t.Fatalf("failed to set key '%s', to value '%s': %s", testKey, testValue2, err)
+	}
+
+	// get test key with value2
+	val2, err := store.Get(ctx, testKey)
+	if err != nil {
+		t.Fatalf("failed to get key '%s': %s", testKey, err)
+	}
+	if !bytes.Equal(testValue2, val2) {
+		t.Fatalf("key='%s' value='%s' doesn't match, expected='%s'", testKey, val2, testValue2)
+	}
+
+	// get a missing key
+	const keyNotExists = "key-not-exists"
+	val3, err := store.Get(ctx, []byte(keyNotExists))
+	if !errors.Is(err, kv.ErrNotFound) {
+		t.Fatalf("get key='%s' err=%s, expected not found", keyNotExists, err)
+	}
+	if val3 != nil {
+		t.Fatalf("get key='%s' value='%s', expected nil", keyNotExists, val3)
+	}
+}
+
+func testStoreDelete(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	t.Run("exists", func(t *testing.T) {
+		keyToDel := []byte("key-to-delete")
+		err := store.Set(ctx, keyToDel, []byte("value to delete"))
+		if err != nil {
+			t.Fatalf("failed to set key='%s': %s", keyToDel, err)
+		}
+		err = store.Delete(ctx, keyToDel)
+		if err != nil {
+			t.Fatalf("failed to delete key='%s': %s", keyToDel, err)
+		}
+	})
+
+	t.Run("non_exists", func(t *testing.T) {
+		keyToDel := []byte("missing-key-to-delete")
+		err := store.Delete(ctx, keyToDel)
+		if !errors.Is(err, kv.ErrNotFound) {
+			t.Fatalf("delete missing key '%s', err=%v expected=%s", keyToDel, err, kv.ErrNotFound)
+		}
+	})
+}
+
+func testStoreSetIf(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	t.Run("previous_value", func(t *testing.T) {
+		// setup initial value
+		key := []byte("setIfKey1")
+		value := []byte("val1")
+		err := store.Set(ctx, key, value)
+		if err != nil {
+			t.Fatalf("failed to set key=%s, value=%s: %s", key, value, err)
+		}
+
+		// should set new value if old value is kept
+		newValue := []byte("val2")
+		err = store.SetIf(ctx, key, newValue, value)
+		if err != nil {
+			t.Fatalf("failed to setIf key=%s value=%s pred=%s: %s", key, newValue, value, err)
+		}
+
+		// should NOT set new value if old value is not kept
+		newValue = []byte("val3")
+		err = store.SetIf(ctx, key, newValue, value)
+		if !errors.Is(err, kv.ErrNotFound) {
```

Comment:
This should use a different error code: it is "condition not satisfied", not "item not found" (which does not make sense for a PUT).

Reply:
I was looking at it as 'not found' as the item with the desired predicate not found.
But @itaiad200 and already pointed this - so I guess most of you prefer different codes to handle - one is key not found, and the second one is for the condition.
Please let me know if I misunderstood - I'll align the implementation.